File: libs/bus-core/src/events/events.ts
export const FollowSubject = 'events.follow'
export const SubscribeSubject = 'events.subscribe'
export const SubGiftSubject = 'events.sub_gift'
export const ReSubscribeSubject = 'events.resubscribe'
export const RedemptionCreatedSubject = 'events.redemption_created'
export const CommandUsedSubject = 'events.command_used'
export const FirstUserMessageSubject = 'events.first_user_message'
export const RaidedSubject = 'events.raided'
export const TitleOrCategoryChangedSubject = 'events.title_or_category_changed'
export const StreamOnlineSubject = 'events.stream_online'
export const StreamOfflineSubject = 'events.stream_offline'
export const ChatClearSubject = 'events.chat_clear'
export const DonateSubject = 'events.donate'
export const KeywordMatchedSubject = 'events.keyword_matched'
export const GreetingSendedSubject = 'events.greeting_sended'
export const PollBeginSubject = 'events.poll_begin'
export const PollProgressSubject = 'events.poll_progress'
export const PollEndSubject = 'events.poll_end'
export const PredictionBeginSubject = 'events.prediction_begin'
export const PredictionProgressSubject = 'events.prediction_progress'
export const PredictionLockSubject = 'events.prediction_lock'
export const PredictionEndSubject = 'events.prediction_end'
export const StreamFirstUserJoinSubject = 'events.stream_first_user_join'
export const ChannelBanSubject = 'events.channel_ban'
export const ChannelUnbanRequestCreateSubject = 'events.channel_unban_request_create'
export const ChannelUnbanRequestResolveSubject = 'events.channel_unban_request_resolve'
export const ChannelMessageDeleteSubject = 'events.channel_message_delete'
export const ModeratorAddedSubject = 'events.moderator_added'
export const ModeratorRemovedSubject = 'events.moderator_removed'
export const VipAddedSubject = 'events.vip_added'
export const VipRemovedSubject = 'events.vip_removed'

export interface BaseInfo {
	channel_id: string
	channel_name: string
}

export interface FollowMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	user_id: string
}

export interface SubscribeMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	level: string
	user_id: string
}

export interface SubGiftMessage {
	base_info: BaseInfo
	sender_user_name: string
	sender_display_name: string
	target_user_name: string
	target_display_name: string
	level: string
	sender_user_id: string
}

export interface ReSubscribeMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	months: number
	streak: number
	is_prime: boolean
	message: string
	level: string
	user_id: string
}

export interface RedemptionCreatedMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	id: string
	reward_name: string
	reward_cost: string
	input?: string
	user_id: string
}

export interface CommandUsedMessage {
	base_info: BaseInfo
	command_id: string
	command_name: string
	user_name: string
	user_display_name: string
	command_input: string
	user_id: string
	is_default: boolean
	default_command_name: string
	message_id: string
}

export interface FirstUserMessageMessage {
	base_info: BaseInfo
	user_id: string
	user_name: string
	user_display_name: string
	message_id: string
}

export interface RaidedMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	viewers: number
	user_id: string
}

export interface TitleOrCategoryChangedMessage {
	base_info: BaseInfo
	old_title: string
	new_title: string
	old_category: string
	new_category: string
}

export interface StreamOnlineMessage {
	base_info: BaseInfo
	title: string
	category: string
}

export interface StreamOfflineMessage {
	base_info: BaseInfo
}

export interface ChatClearMessage {
	base_info: BaseInfo
}

export interface DonateMessage {
	base_info: BaseInfo
	user_name: string
	amount: string
	currency: string
	message: string
}

export interface KeywordMatchedMessage {
	base_info: BaseInfo
	keyword_id: string
	keyword_name: string
	keyword_response: string
	user_id: string
	user_name: string
	user_display_name: string
}

export interface GreetingSendedMessage {
	base_info: BaseInfo
	user_id: string
	user_name: string
	user_display_name: string
	greeting_text: string
}

export interface PollChoice {
	id: string
	title: string
	bits_votes: number
	channels_points_votes: number
	votes: number
}

export interface PollBitsVotes {
	enabled: boolean
	amount_per_vote: number
}

export interface PollChannelPointsVotes {
	enabled: boolean
	amount_per_vote: number
}

export interface PollInfo {
	title: string
	choices: PollChoice[]
	bits_voting: PollBitsVotes
	channels_points_voting: PollChannelPointsVotes
}

export interface PollBeginMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	info: PollInfo
}

export interface PollProgressMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	info: PollInfo
}

export interface PollEndMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	info: PollInfo
}

export interface PredictionTopPredictor {
	user_name: string
	user_display_name: string
	user_id: string
	points_used: number
	points_win?: number
}

export interface PredictionOutcome {
	id: string
	title: string
	color: string
	users: number
	channel_points: number
	top_predictors: PredictionTopPredictor[]
}

export interface PredictionInfo {
	title: string
	outcomes: PredictionOutcome[]
}

export interface PredictionBeginMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	info: PredictionInfo
}

export interface PredictionProgressMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	info: PredictionInfo
}

export interface PredictionLockMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	info: PredictionInfo
}

export interface PredictionEndMessage {
	base_info: BaseInfo
	user_name: string
	user_display_name: string
	info: PredictionInfo
	winning_outcome_id: string
}

export interface StreamFirstUserJoinMessage {
	base_info: BaseInfo
	user_name: string
}

export interface ChannelBanMessage {
	base_info: BaseInfo
	user_name: string
	user_login: string
	broadcaster_user_name: string
	broadcaster_user_login: string
	moderator_user_name: string
	moderator_user_login: string
	reason: string
	ends_at: string
	is_permanent: boolean
}

export interface ChannelUnbanRequestCreateMessage {
	base_info: BaseInfo
	user_name: string
	user_login: string
	broadcaster_user_name: string
	broadcaster_user_login: string
	text: string
}

export interface ChannelUnbanRequestResolveMessage {
	base_info: BaseInfo
	user_name: string
	user_login: string
	broadcaster_user_name: string
	broadcaster_user_login: string
	moderator_user_name: string
	moderator_user_login: string
}

export interface ChannelMessageDeleteMessage {
	base_info: BaseInfo
	user_id: string
	user_name: string
	user_login: string
	broadcaster_user_name: string
	broadcaster_user_login: string
	message_id: string
}

export interface ModeratorAddedMessage {
	base_info: BaseInfo
	user_id: string
	user_name: string
}

export interface ModeratorRemovedMessage {
	base_info: BaseInfo
	user_id: string
	user_name: string
}

export interface VipAddedMessage {
	base_info: BaseInfo
	user_id: string
	user_name: string
}

export interface VipRemovedMessage {
	base_info: BaseInfo
	user_id: string
	user_name: string
}

File: libs/bus-core/events/events.go
package events

const (
	FollowSubject                     = "events.follow"
	SubscribeSubject                  = "events.subscribe"
	SubGiftSubject                    = "events.sub_gift"
	ReSubscribeSubject                = "events.resubscribe"
	RedemptionCreatedSubject          = "events.redemption_created"
	CommandUsedSubject                = "events.command_used"
	FirstUserMessageSubject           = "events.first_user_message"
	RaidedSubject                     = "events.raided"
	TitleOrCategoryChangedSubject     = "events.title_or_category_changed"
	StreamOnlineSubject               = "events.stream_online"
	StreamOfflineSubject              = "events.stream_offline"
	ChatClearSubject                  = "events.chat_clear"
	DonateSubject                     = "events.donate"
	KeywordMatchedSubject             = "events.keyword_matched"
	GreetingSendedSubject             = "events.greeting_sended"
	PollBeginSubject                  = "events.poll_begin"
	PollProgressSubject               = "events.poll_progress"
	PollEndSubject                    = "events.poll_end"
	PredictionBeginSubject            = "events.prediction_begin"
	PredictionProgressSubject         = "events.prediction_progress"
	PredictionLockSubject             = "events.prediction_lock"
	PredictionEndSubject              = "events.prediction_end"
	StreamFirstUserJoinSubject        = "events.stream_first_user_join"
	ChannelBanSubject                 = "events.channel_ban"
	ChannelUnbanRequestCreateSubject  = "events.channel_unban_request_create"
	ChannelUnbanRequestResolveSubject = "events.channel_unban_request_resolve"
	ChannelMessageDeleteSubject       = "events.channel_message_delete"
	VipAddedSubject                   = "events.vip_added"
	VipRemovedSubject                 = "events.vip_removed"
	ModeratorAddedSubject             = "events.moderator_added"
	ModeratorRemovedSubject           = "events.moderator_removed"
	ChannelUnbanSubject               = "events.channel_unban"
)

type BaseInfo struct {
	ChannelID   string `json:"channel_id"`
	ChannelName string `json:"channel_name"`
}

type FollowMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	UserID          string   `json:"user_id"`
}

type SubscribeMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	Level           string   `json:"level"`
	UserID          string   `json:"user_id"`
}

type SubGiftMessage struct {
	BaseInfo          BaseInfo `json:"base_info"`
	SenderUserName    string   `json:"sender_user_name"`
	SenderDisplayName string   `json:"sender_display_name"`
	TargetUserName    string   `json:"target_user_name"`
	TargetDisplayName string   `json:"target_display_name"`
	Level             string   `json:"level"`
	SenderUserID      string   `json:"sender_user_id"`
}

type ReSubscribeMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	Months          int64    `json:"months"`
	Streak          int64    `json:"streak"`
	IsPrime         bool     `json:"is_prime"`
	Message         string   `json:"message"`
	Level           string   `json:"level"`
	UserID          string   `json:"user_id"`
}

type RedemptionCreatedMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	ID              string   `json:"id"`
	RewardName      string   `json:"reward_name"`
	RewardCost      string   `json:"reward_cost"`
	Input           *string  `json:"input,omitempty"`
	UserID          string   `json:"user_id"`
}

type CommandUsedMessage struct {
	BaseInfo           BaseInfo `json:"base_info"`
	CommandID          string   `json:"command_id"`
	CommandName        string   `json:"command_name"`
	UserName           string   `json:"user_name"`
	UserDisplayName    string   `json:"user_display_name"`
	CommandInput       string   `json:"command_input"`
	UserID             string   `json:"user_id"`
	IsDefault          bool     `json:"is_default"`
	DefaultCommandName *string  `json:"default_command_name"`
	MessageID          string   `json:"message_id"`
}

type FirstUserMessageMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserID          string   `json:"user_id"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	MessageID       string   `json:"message_id"`
}

type RaidedMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	Viewers         int64    `json:"viewers"`
	UserID          string   `json:"user_id"`
}

type TitleOrCategoryChangedMessage struct {
	BaseInfo    BaseInfo `json:"base_info"`
	OldTitle    string   `json:"old_title"`
	NewTitle    string   `json:"new_title"`
	OldCategory string   `json:"old_category"`
	NewCategory string   `json:"new_category"`
}

type ChatClearMessage struct {
	BaseInfo BaseInfo `json:"base_info"`
}

type DonateMessage struct {
	BaseInfo BaseInfo `json:"base_info"`
	UserName string   `json:"user_name"`
	Amount   string   `json:"amount"`
	Currency string   `json:"currency"`
	Message  string   `json:"message"`
}

type KeywordMatchedMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	KeywordID       string   `json:"keyword_id"`
	KeywordName     string   `json:"keyword_name"`
	KeywordResponse string   `json:"keyword_response"`
	UserID          string   `json:"user_id"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
}

type GreetingSendedMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserID          string   `json:"user_id"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	GreetingText    string   `json:"greeting_text"`
}

type PollChoice struct {
	ID                  string `json:"id"`
	Title               string `json:"title"`
	BitsVotes           uint64 `json:"bits_votes"`
	ChannelsPointsVotes uint64 `json:"channels_points_votes"`
	Votes               uint64 `json:"votes"`
}

type PollBitsVotes struct {
	Enabled       bool   `json:"enabled"`
	AmountPerVote uint64 `json:"amount_per_vote"`
}

type PollChannelPointsVotes struct {
	Enabled       bool   `json:"enabled"`
	AmountPerVote uint64 `json:"amount_per_vote"`
}

type PollInfo struct {
	Title                string                 `json:"title"`
	Choices              []PollChoice           `json:"choices"`
	BitsVoting           PollBitsVotes          `json:"bits_voting"`
	ChannelsPointsVoting PollChannelPointsVotes `json:"channels_points_voting"`
}

type PollBeginMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	Info            PollInfo `json:"info"`
}

type PollProgressMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	Info            PollInfo `json:"info"`
}

type PollEndMessage struct {
	BaseInfo        BaseInfo `json:"base_info"`
	UserName        string   `json:"user_name"`
	UserDisplayName string   `json:"user_display_name"`
	Info            PollInfo `json:"info"`
}

type PredictionTopPredictor struct {
	UserName        string  `json:"user_name"`
	UserDisplayName string  `json:"user_display_name"`
	UserID          string  `json:"user_id"`
	PointsUsed      uint64  `json:"points_used"`
	PointsWin       *uint64 `json:"points_win,omitempty"`
}

type PredictionOutcome struct {
	ID            string                   `json:"id"`
	Title         string                   `json:"title"`
	Color         string                   `json:"color"`
	Users         uint64                   `json:"users"`
	ChannelPoints uint64                   `json:"channel_points"`
	TopPredictors []PredictionTopPredictor `json:"top_predictors"`
}

type PredictionInfo struct {
	Title    string              `json:"title"`
	Outcomes []PredictionOutcome `json:"outcomes"`
}

type PredictionBeginMessage struct {
	BaseInfo        BaseInfo       `json:"base_info"`
	UserName        string         `json:"user_name"`
	UserDisplayName string         `json:"user_display_name"`
	Info            PredictionInfo `json:"info"`
}

type PredictionProgressMessage struct {
	BaseInfo        BaseInfo       `json:"base_info"`
	UserName        string         `json:"user_name"`
	UserDisplayName string         `json:"user_display_name"`
	Info            PredictionInfo `json:"info"`
}

type PredictionLockMessage struct {
	BaseInfo        BaseInfo       `json:"base_info"`
	UserName        string         `json:"user_name"`
	UserDisplayName string         `json:"user_display_name"`
	Info            PredictionInfo `json:"info"`
}

type PredictionEndMessage struct {
	BaseInfo         BaseInfo       `json:"base_info"`
	UserName         string         `json:"user_name"`
	UserDisplayName  string         `json:"user_display_name"`
	Info             PredictionInfo `json:"info"`
	WinningOutcomeID string         `json:"winning_outcome_id"`
}

type StreamFirstUserJoinMessage struct {
	BaseInfo  BaseInfo `json:"base_info"`
	UserID    string   `json:"user_id"`
	UserLogin string   `json:"user_login"`
}

type ChannelBanMessage struct {
	BaseInfo             BaseInfo `json:"base_info"`
	UserID               string   `json:"user_id"`
	UserName             string   `json:"user_name"`
	UserLogin            string   `json:"user_login"`
	BroadcasterUserName  string   `json:"broadcaster_user_name"`
	BroadcasterUserLogin string   `json:"broadcaster_user_login"`
	ModeratorUserID      string   `json:"moderator_id"`
	ModeratorUserName    string   `json:"moderator_user_name"`
	ModeratorUserLogin   string   `json:"moderator_user_login"`
	Reason               string   `json:"reason"`
	EndsAt               string   `json:"ends_at"`
	IsPermanent          bool     `json:"is_permanent"`
}

type ChannelUnbanMessage struct {
	BaseInfo             BaseInfo `json:"base_info"`
	UserID               string   `json:"user_id"`
	UserName             string   `json:"user_name"`
	UserLogin            string   `json:"user_login"`
	BroadcasterUserName  string   `json:"broadcaster_user_name"`
	BroadcasterUserLogin string   `json:"broadcaster_user_login"`
	ModeratorUserID      string   `json:"moderator_id"`
	ModeratorUserName    string   `json:"moderator_user_name"`
	ModeratorUserLogin   string   `json:"moderator_user_login"`
}

type ChannelUnbanRequestCreateMessage struct {
	BaseInfo             BaseInfo `json:"base_info"`
	UserName             string   `json:"user_name"`
	UserLogin            string   `json:"user_login"`
	BroadcasterUserName  string   `json:"broadcaster_user_name"`
	BroadcasterUserLogin string   `json:"broadcaster_user_login"`
	Text                 string   `json:"text"`
}

type ChannelUnbanRequestResolveMessage struct {
	BaseInfo             BaseInfo `json:"base_info"`
	UserID               string   `json:"user_id"`
	UserName             string   `json:"user_name"`
	UserLogin            string   `json:"user_login"`
	BroadcasterUserName  string   `json:"broadcaster_user_name"`
	BroadcasterUserLogin string   `json:"broadcaster_user_login"`
	ModeratorUserID      string   `json:"moderator_id"`
	ModeratorUserName    string   `json:"moderator_user_name"`
	ModeratorUserLogin   string   `json:"moderator_user_login"`
	Declined             bool     `json:"declined"`
	Reason               string   `json:"reason"`
}

type ChannelMessageDeleteMessage struct {
	BaseInfo             BaseInfo `json:"base_info"`
	UserId               string   `json:"user_id"`
	UserName             string   `json:"user_name"`
	UserLogin            string   `json:"user_login"`
	BroadcasterUserName  string   `json:"broadcaster_user_name"`
	BroadcasterUserLogin string   `json:"broadcaster_user_login"`
	MessageId            string   `json:"message_id"`
}

type VipAddedMessage struct {
	BaseInfo BaseInfo `json:"base_info"`
	UserName string   `json:"user_name"`
	UserID   string   `json:"user_id"`
}

type VipRemovedMessage struct {
	BaseInfo BaseInfo `json:"base_info"`
	UserName string   `json:"user_name"`
	UserID   string   `json:"user_id"`
}

type ModeratorAddedMessage struct {
	BaseInfo BaseInfo `json:"base_info"`
	UserName string   `json:"user_name"`
	UserID   string   `json:"user_id"`
}

type ModeratorRemovedMessage struct {
	BaseInfo BaseInfo `json:"base_info"`
	UserName string   `json:"user_name"`
	UserID   string   `json:"user_id"`
}
